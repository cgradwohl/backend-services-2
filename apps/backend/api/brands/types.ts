import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiBrandsGetResponse,
  IApiBrandsListResponse,
  IApiBrandsPostRequest,
  IApiBrandsPostResponse,
  IApiBrandsPutRequest,
  IApiBrandsPutResponse,
} from "~/types.public";

export type ApiBrandsResponse =
  | IApiBrandsGetResponse
  | IApiBrandsListResponse
  | IApiBrandsPostResponse
  | IApiBrandsPutResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsListResponse;
}>;

export type PostRequestBody = IApiBrandsPostRequest;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPostResponse;
}>;

export type PutRequestBody = IApiBrandsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
