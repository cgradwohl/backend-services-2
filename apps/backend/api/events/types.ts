import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiEventsGetResponse,
  IApiEventsListResponse,
  IApiEventsPutRequest,
  IApiEventsPutResponse,
} from "~/types.public";

export type ApiEventsResponse = IApiEventsGetResponse | IApiEventsListResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsListResponse;
}>;

export type PutRequestBody = IApiEventsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
