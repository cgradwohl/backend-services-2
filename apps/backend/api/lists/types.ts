import { ApiRequestContext, Response } from "~/lib/lambda-response";

import {
  IApiDeleteListItemResponse,
  IApiGetListItemResponse,
  IApiGetListItemsResponse,
  IApiGetListItemSubscriptionsResponse,
  IApiListItemSubscribeResponse,
  IApiListItemUnsubscribeResponse,
  IApiPutListItemResponse,
  IApiRestoreListItemResponse,
} from "~/types.public";

export type HttpEventHandler<T extends Response<any>> = (
  context: ApiRequestContext
) => Promise<T>;

export type ListApiResponse =
  | IApiDeleteListItemResponse
  | IApiGetListItemResponse
  | IApiGetListItemsResponse
  | IApiGetListItemSubscriptionsResponse
  | IApiListItemSubscribeResponse
  | IApiListItemUnsubscribeResponse
  | IApiPutListItemResponse
  | IApiRestoreListItemResponse;
