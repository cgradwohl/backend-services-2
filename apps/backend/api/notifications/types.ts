import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiNotificationGetResponse,
  IApiNotificationGetSubmissionChecksResponse,
  IApiNotificationListResponse,
  IApiNotificationPutSubmissionChecksResponse,
} from "~/types.public";

export type ApiNotificationResponse =
  | IApiNotificationGetResponse
  | IApiNotificationListResponse
  | IApiNotificationGetSubmissionChecksResponse
  | IApiNotificationPutSubmissionChecksResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationGetResponse;
}>;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  status: number;
}>;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  status: number;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationListResponse;
}>;

export type GetSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationGetSubmissionChecksResponse;
}>;

export type PutSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationPutSubmissionChecksResponse;
}>;

export type DeleteSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204 | 409;
}>;
