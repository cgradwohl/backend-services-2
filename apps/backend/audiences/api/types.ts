import { ApiRequestContext } from "~/lib/lambda-response";
import { FilterConfig } from "~/audiences/stores/dynamo/types";

export type Resource =
  | "/audiences"
  | "/audiences/{audience_id}"
  | "/audiences/{audience_id}/members"
  | "/audiences/{audience_id}/members/{member_id}";

export type MemberResource = "/members/{member_id}/audiences";

export type BaseAudienceResponse = {};

export interface AudienceHandler<T extends BaseAudienceResponse> {
  (context: ApiRequestContext): Promise<T>;
}

export type Audience = {
  created_at: string;
  description: string;
  id: string;
  name: string;
  // TODO: https://linear.app/trycourier/issue/C-5533/rule-engine-ux-for-allowing-people-to-create-audiences-rule
  // rules will never be null, but until we have a UI for creating audiences, we need to allow it to be null
  filter: FilterConfig | null;
  updated_at: string;
};

export type AudienceMember = {
  added_at: string;
  audience_id: string;
  audience_version: number;
  member_id: string;
  reason: string;
};

export interface AudienceListResponse extends BaseAudienceResponse {
  body: {
    paging: {
      cursor: string;
      more: boolean;
    };
    items: Audience[];
  };
}

export interface AudienceDeleteResponse extends BaseAudienceResponse {
  status: 204;
}

export interface AudienceGetResponse extends BaseAudienceResponse {
  status: 200;
  body: Audience;
}

export interface AudienceMemberGetResponse extends BaseAudienceResponse {
  status: 200;
  body: AudienceMember;
}

export interface AudienceMemberListResponse extends BaseAudienceResponse {
  status: 200;
  body: {
    items: AudienceMember[];
    paging: {
      cursor: string;
      more: boolean;
    };
  };
}

export interface AudiencePutResponse extends BaseAudienceResponse {
  status: 200;
  body: {
    audience: Audience;
  };
}

export interface AudienceMemberPutResponse extends BaseAudienceResponse {
  status: 200;
  body: {
    audience_member: AudienceMember;
  };
}

export type Method = "get" | "put" | "delete";

export type UserMethod = "get";
