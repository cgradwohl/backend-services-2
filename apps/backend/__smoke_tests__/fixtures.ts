const _emailEvents = (() => {
  if (process.env.SMOKE_TEST_EVENT) {
    return [process.env.SMOKE_TEST_EVENT];
  }

  if (process.env.STAGE === "staging") {
    return ["SAPBJ1JWR44V2MQ3NBRFRVAC594W"];
  }
  if (process.env.STAGE === "production") {
    return ["14171CNHXG4JHXNH2NRMBYJ0Y20P"];
  }

  return [];
})();

export const emailEvents = _emailEvents;
