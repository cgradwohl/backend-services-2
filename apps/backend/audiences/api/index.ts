import { handleApi } from "~/lib/lambda-response";
import { getAudiences } from "./get";
import * as AudienceTypes from "./types";
import { deleteAudience, getAudience, putAudience } from "./{audience_id}";
import {
  getAudienceMember,
  getAudienceMembers,
  putAudienceMember,
} from "./{audience_id}/members";
const handlers: {
  [Resource in AudienceTypes.Resource]: Partial<{
    [Method in AudienceTypes.Method]: AudienceTypes.AudienceHandler<AudienceTypes.BaseAudienceResponse>;
  }>;
} = {
  "/audiences": {
    get: getAudiences,
  },
  "/audiences/{audience_id}": {
    delete: deleteAudience,
    get: getAudience,
    put: putAudience,
  },
  "/audiences/{audience_id}/members": {
    get: getAudienceMembers,
  },
  "/audiences/{audience_id}/members/{member_id}": {
    get: getAudienceMember,
    put: putAudienceMember,
  },
};

export const handler = handleApi<AudienceTypes.BaseAudienceResponse>(
  async (context) => {
    const method =
      context.event.httpMethod.toLowerCase() as AudienceTypes.Method;
    const resource = context.event.resource as AudienceTypes.Resource;
    const apiHandler = handlers[resource]?.[method];
    const response = await apiHandler?.(context);
    return {
      ...response,
    };
  }
);
