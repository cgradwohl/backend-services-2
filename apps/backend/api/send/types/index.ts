import {
  IBrandColors,
  IBrandSettingsEmail,
  IBrandSettingsInApp,
  IBrandSnippets,
} from "~/lib/brands/types";
import { IBrand } from "~/types.api";
import { IProfilePreferences } from "~/types.public";
import { ElementalContent } from "./courier-elemental";

export * from "./courier-elemental";

export interface MessageData extends Record<string, any> {}

type InvalidListRecipient = {
  user_id: string;
  list_pattern: string;
};

type ListRecipientType = Record<string, unknown> & {
  [key in keyof InvalidListRecipient]?: never;
};
export interface ListRecipient extends ListRecipientType {
  list_id?: string;
  data?: MessageData;
}

export interface AudienceRecipient {
  audience_id: string;
  data?: MessageData;
}

type InvalidListPatternRecipient = {
  user_id: string;
  list_id: string;
};

type ListPatternRecipientType = Record<string, unknown> & {
  [key in keyof InvalidListPatternRecipient]?: never;
};
export interface ListPatternRecipient extends ListPatternRecipientType {
  list_pattern?: string;
  data?: MessageData;
}

type InvalidUserRecipient = {
  list_id: string;
  list_pattern: string;
};

type UserRecipientType = Record<string, unknown> & {
  [key in keyof InvalidUserRecipient]?: never;
};

export interface UserRecipient extends UserRecipientType {
  data?: MessageData;
  email?: string;
  locale?: string;
  user_id?: string;
  phone_number?: string;
  preferences?: IProfilePreferences;
}

export type Recipient =
  | AudienceRecipient
  | ListRecipient
  | ListPatternRecipient
  | UserRecipient;

export type MessageRecipient = Recipient | Recipient[];

export interface ElementalContentSugar {
  body?: string;
  title?: string;
}

export interface Timeout {
  /** I recommend deprecating the record format, makes overrides a bit confusing */
  provider?: number | { [provider: string]: number };
  /** I recommend deprecating the record format, makes overrides a bit confusing */
  channel?: number | { [provider: string]: number };
  message?: number;
}

/** @deprecated */
export interface TimeoutDate<T> {
  message: T;
  channel?: T;
  channels?: {
    [channel: string]: ChannelTimeout<T>;
  };
  provider?: T;
  providers?: { [provider: string]: T };
}

/** @deprecated Timeout specified as an epoch date (seconds after unix epoch) */
export interface TimeoutDateEpochSeconds extends TimeoutDate<number> {}

/** @deprecated Timeout specified as ISO 8601 date string  */
export interface TimeoutDateIso8601 extends TimeoutDate<string> {}

export interface ChannelTimeout<T> {
  channel?: T;
  provider?: T;
}

export type Content = ElementalContentSugar | ElementalContent;

export interface MessageDelayDuration {
  duration: string; // friendly ("1 hour"), milliseconds, ISO-8601 duration
  until?: never;
}

export interface MessageDelayUntil {
  duration?: never;
  until: string; // friendly ("Thursday at 10am"), Unix Epoch, ISO-8601 date-stamp w/ TZ
}

export type MessageDelay = MessageDelayDuration | MessageDelayUntil;

export interface MessageBrandV2 {
  version: "2022-05-17";
  colors?: {
    primary?: string;
    secondary?: string;
    tertiary?: string;
  };
  logo?: {
    href?: string;
    image: string;
  };
  locales?: {
    [key: string]: Omit<MessageBrand, "locales" | "version">;
  };
}

export type MessageBrandV1 = Partial<Omit<IBrand, "version">> & {
  settings: IBrand["settings"];
  version?: "2020-06-19T18:51:36.083Z" | "2020-06-19";
};

export type MessageBrand = MessageBrandV1 | MessageBrandV2;

export interface BaseMessage {
  brand_id?: string;
  brand?: MessageBrand;
  channels?: MessageChannels;
  data?: MessageData;
  delay?: MessageDelay;
  metadata?: MessageMetadata;
  providers?: MessageProviders;
  routing?: MessageRouting;
  timeout?: Timeout;
  to: MessageRecipient;
}

interface TrackingOverride {
  open: boolean;
}
export interface MessageChannelEmailOverride {
  attachments?: Record<string, any>[];
  bcc?: string;
  // content?: Content; TODO: https://linear.app/trycourier/issue/C-4462/add-content-support-to-channel-overrides
  brand?: {
    snippets?: IBrandSnippets;
    settings?: {
      colors?: IBrandColors;
      email?: IBrandSettingsEmail;
    };
  };
  cc?: string;
  from?: string;
  html?: string;
  reply_to?: string;
  subject?: string;
  text?: string;
  tracking?: TrackingOverride;
}

export interface MessageChannelPushOverride {
  body?: string;
  brand?: {
    snippets?: IBrandSnippets;
    settings?: {
      colors?: IBrandColors;
      inapp?: IBrandSettingsInApp;
    };
  };
  // content?: Content; TODO: https://linear.app/trycourier/issue/C-4462/add-content-support-to-channel-overrides
  click_action?: string;
  data?: Record<string, any>;
  icon?: string;
  reply_to?: undefined;
  title?: string;
}

export interface MessageProviderConfig {
  override?: Record<string, any>;
  if?: string;
  timeout?: number;
  metadata?: {
    utm?: UTM;
  };
}

export interface MessageProviders {
  [provider: string]: MessageProviderConfig;
}

export interface MessageChannelConfig {
  brand_id?: string; // TODO: pick this up in prepare (like override)?
  providers?: string[];
  routing_method?: "all" | "single";
  if?: string;
  timeout?: number | ChannelTimeout<number>;
  metadata?: {
    utm?: UTM;
  };
  override?: MessageChannelEmailOverride | MessageChannelPushOverride;
}
export interface MessageChannels {
  [channel: string]: MessageChannelConfig;
}

export type MessageRoutingMethod = "all" | "single";
export type MessageRoutingChannel = string | MessageRouting;
export interface MessageRouting {
  method: MessageRoutingMethod;
  channels: MessageRoutingChannel[];
}

export interface UTM {
  source?: string;
  medium?: string;
  campaign?: string;
  term?: string;
  content?: string;
}

export interface UTMMap {
  message?: UTM;
  channels?: {
    [channel: string]: UTM;
  };
  providers?: {
    [provider: string]: UTM;
  };
}

export interface MessageMetadata {
  event?: string;
  tags?: string[];
  utm?: UTM;
  trace_id?: string;
}

export interface ContentMessage extends BaseMessage {
  content: Content;
}

export interface TemplateMessage extends BaseMessage {
  template: string;
}

export type Message = ContentMessage | TemplateMessage;

export enum SequenceActions {
  cancel = "cancel",
  emit = "emit",
  fetchData = "fetch-data",
  send = "send",
}

export interface SendSequenceAction {
  action: SequenceActions;
  message: Message;
}

export type Sequence = Array<SendSequenceAction>;
// | EmitSequenceAction
// | FetchDataSequenceAction
// | CancelSequenceAction>;

export interface MessageRequest {
  message: Message;
  sequence?: never;
}

export interface SequenceRequest {
  message?: never;
  sequence: Sequence;
}

export type RequestV2 = SequenceRequest | MessageRequest;
