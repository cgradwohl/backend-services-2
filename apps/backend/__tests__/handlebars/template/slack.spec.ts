import getProviderTemplate from "~/handlebars/template";
import { TemplateConfig } from "~/handlebars/template/types";
import hydrateBlock from "~/lib/blocks/hydrate-slate-block";
import createLinkHandler from "~/lib/link-handler";
import createVariableHandler from "~/lib/variable-handler";
import { ITextBlock } from "~/types.api";

import fixtures from "../__fixtures__";
import mockData from "./util/mock-data";
import mockProfile from "./util/mock-profile";

const { blocks: blockFixtures, slate: slateFixtures } = fixtures;

let usingTemplate = false;
jest.mock("~/handlebars/partials", () => {
  const partials = jest.requireActual("~/handlebars/partials");
  const courierTemplate = partials.default.email["courier-template"];

  // going to dynamically render the template based on usingTemplate
  Object.defineProperty(partials.default.email, "courier-template", {
    get: () => (usingTemplate ? courierTemplate : undefined),
  });

  return partials;
});

const hydrateBlocks = (blocks: any[]) => blocks.map(hydrateBlock);

describe("slack template blocks", () => {
  const testBlocks = [
    "textBlockHeader",
    "textBlock",
    "actionBlock",
    "actionBlockLink",
    "actionBlockWebhook",
    "conditionalTextBlock",
    "conditionalTextBlockShown",
    "dividerBlock",
    "dividerBlockWithColor",
    "imageBlock",
    "imageBlockWithVars",
    "imageBlockWithFullAlign",
    "jsonnetBlock",
    "jsonnetBlockArray",
    "listBlock",
    "listBlockBulletsEmpty",
    "listBlockWithChildren",
    "listBlockBullets",
    "listBlockBulletsWithChildren",
    "listBlockTransparent",
    "listBlockTransparentWithChildren",
    "listBlockTransparentWithTopImage",
    "listBlockTransparentWithChildImage",
    "listBlockTransparentWithChildNotFound",
    "listBlockUsingObjects",
    "markdownBlock",
    "markdownBlockMultiLine",
    "markdownBlockWithBoldVariable",
    "markdownBlockWithTextWithTags",
    "markdownBlockWithVariableWithTags",
    "quoteBlockWithLink",
    "templateBlock",
    "textBlockWithLink",
    "textBlockWithMarks",
    "textBlockWithVariable",
    "textBlockWithVariableEscaping",
    "textBlockWithVariableWithLineReturns",
    "textBlockWithVariableWithWindowsLineReturns",
    "textBlockWhitespace",
    "textBlockWithLinkWithWhitespace",
    "welcomeBlocks",
  ];

  const only = false; // "jsonnetBlock";

  beforeEach(() => (usingTemplate = false));

  testBlocks.forEach((block) => {
    const fn = only && only === block ? it.only : it;

    fn(`should render ${block}`, () => {
      const allBlocks = hydrateBlocks(blockFixtures[block]);
      const channelBlockIds = allBlocks.map((block) => block.id);
      const config: TemplateConfig = { email: { emailTemplateConfig: {} } };
      const template = getProviderTemplate({
        allBlocks,
        channelBlockIds,
        config,
        provider: "slack",
      });
      const slackTemplate = template.slackRenderer();
      const data = {
        data: mockData,
        event: "mockEvent",
        profile: mockProfile,
        recipient: "mockRecipient",
        urls: { opened: "https://abc.ct0.app/o/open-tracking" },
      };
      const rendered = slackTemplate.render(
        createVariableHandler({ value: data }).getScoped("data"),
        createLinkHandler({})
      );
      expect(rendered).toMatchSnapshot();
    });
  });
});

describe("slack slate", () => {
  const testSlate = [
    "slateBoldItalicUnderline",
    "slateItalicAroundBoldSlate",
    "slateLink",
    "slateLinkWithVars",
    "slateNewLines",
    "slateScaryChars",
    "slateVariable",
  ];

  const only = false; // "slateBoldItalicUnderline";

  const textBlock = hydrateBlocks(blockFixtures.textBlock)[0] as ITextBlock;

  beforeEach(() => (usingTemplate = false));

  testSlate.forEach((templateName) => {
    const fn = only && only === templateName ? it.only : it;

    fn(`should render ${templateName}`, () => {
      const slateTextBlock: ITextBlock = {
        ...textBlock,
        config: {
          ...textBlock.config,
          value: slateFixtures[templateName],
        },
      };
      const allBlocks = [slateTextBlock];
      const channelBlockIds = [slateTextBlock.id];
      const config: TemplateConfig = {};
      const template = getProviderTemplate({
        allBlocks,
        channelBlockIds,
        config,
      });
      const slackTemplate = template.slackRenderer();
      const data = {
        data: mockData,
        profile: mockProfile,
        urls: { opened: "https://abc.ct0.app/o/open-tracking" },
      };
      const rendered = slackTemplate.render(
        createVariableHandler({ value: data }).getScoped("data"),
        createLinkHandler({})
      );
      expect(rendered).toMatchSnapshot();
    });
  });
});
