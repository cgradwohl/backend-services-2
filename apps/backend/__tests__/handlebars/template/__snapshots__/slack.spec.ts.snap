// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slack slate should render slateBoldItalicUnderline 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Some *bold*, _italic_, and underlined text.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack slate should render slateItalicAroundBoldSlate 1`] = `
Array [
  Object {
    "text": Object {
      "text": "This _is_ *_a_* test",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack slate should render slateLink 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Click <https://example.com/link|this link> to learn more.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack slate should render slateLinkWithVars 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Link to <https://{domain}/{user}/|{firstname}'s {project}>",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack slate should render slateNewLines 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Line 1
Line 2

Line 4",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack slate should render slateScaryChars 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Scary characters are \\", ', <, >, &, \`, =, {, and }",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack slate should render slateVariable 1`] = `
Array [
  Object {
    "text": Object {
      "text": "This is a {variable}.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render actionBlock 1`] = `
Array [
  Object {
    "elements": Array [
      Object {
        "action_id": undefined,
        "text": Object {
          "emoji": true,
          "text": "Click Here",
          "type": "plain_text",
        },
        "type": "button",
        "url": "https://www.courier.com",
      },
    ],
    "type": "actions",
  },
]
`;

exports[`slack template blocks should render actionBlockLink 1`] = `
Array [
  Object {
    "text": Object {
      "text": "<https://www.example.com|link text>",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render actionBlockWebhook 1`] = `
Array [
  Object {
    "elements": Array [
      Object {
        "action_id": "my-action-id",
        "text": Object {
          "emoji": true,
          "text": "Click Here",
          "type": "plain_text",
        },
        "type": "button",
        "url": undefined,
      },
    ],
    "type": "actions",
  },
]
`;

exports[`slack template blocks should render conditionalTextBlock 1`] = `Array []`;

exports[`slack template blocks should render conditionalTextBlockShown 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Conditional text.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render dividerBlock 1`] = `
Array [
  Object {
    "type": "divider",
  },
]
`;

exports[`slack template blocks should render dividerBlockWithColor 1`] = `
Array [
  Object {
    "type": "divider",
  },
]
`;

exports[`slack template blocks should render imageBlock 1`] = `
Array [
  Object {
    "alt_text": "mountains",
    "image_url": "https://s3.amazonaws.com/backend-dev-librarybucket-a6n6jb55kc98/118c6491-a35d-4b81-b10b-d22f7d20da0c/1590005909939_green-mountains.jpg",
    "type": "image",
  },
]
`;

exports[`slack template blocks should render imageBlockWithFullAlign 1`] = `
Array [
  Object {
    "alt_text": " ",
    "image_url": "https://example.com/img.png",
    "type": "image",
  },
]
`;

exports[`slack template blocks should render imageBlockWithVars 1`] = `
Array [
  Object {
    "alt_text": "Description: {imageDescription}",
    "image_url": "http://example.com/{imagePath}",
    "type": "image",
  },
]
`;

exports[`slack template blocks should render jsonnetBlock 1`] = `
Array [
  Object {
    "accessory": Object {
      "alt_text": "Luke Skywalker image",
      "image_url": "https://example.com/luke.png",
      "type": "image",
    },
    "text": Object {
      "text": "Luke Skywalker",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render jsonnetBlockArray 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Number 1",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "text": Object {
      "text": "Number 2",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "text": Object {
      "text": "Number 3",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlock 1`] = `
Array [
  Object {
    "text": Object {
      "text": "• Luke Skywalker
• Boba Fett
• Han Solo",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockBullets 1`] = `
Array [
  Object {
    "text": Object {
      "text": "• Luke Skywalker
• Boba Fett
• Han Solo",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockBulletsEmpty 1`] = `Array []`;

exports[`slack template blocks should render listBlockBulletsWithChildren 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Luke Skywalker
• lightsaber

Boba Fett
• blaster
• Z-6 Jetpack
• flame thrower

Han Solo
• blaster pistol",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockTransparent 1`] = `
Array [
  Object {
    "text": Object {
      "text": "• Luke Skywalker
• Boba Fett
• Han Solo",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockTransparentWithChildImage 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Luke Skywalker
• lightsaber

Boba Fett
• blaster
• Z-6 Jetpack
• flame thrower

Han Solo
• blaster pistol",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockTransparentWithChildNotFound 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Luke Skywalker

Boba Fett

Han Solo",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockTransparentWithChildren 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Luke Skywalker
• lightsaber

Boba Fett
• blaster
• Z-6 Jetpack
• flame thrower

Han Solo
• blaster pistol",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockTransparentWithTopImage 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Luke Skywalker
• lightsaber

Boba Fett
• blaster
• Z-6 Jetpack
• flame thrower

Han Solo
• blaster pistol",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockUsingObjects 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Parent object
• Child object",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render listBlockWithChildren 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Luke Skywalker
• lightsaber

Boba Fett
• blaster
• Z-6 Jetpack
• flame thrower

Han Solo
• blaster pistol",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render markdownBlock 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Hello Luke. Get started with <http://daringfireball.net/projects/markdown/|markdown>.
",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render markdownBlockMultiLine 1`] = `
Array [
  Object {
    "text": Object {
      "text": "​*This is a test*​

The following is a list:

•   A
•   B
•   C
•   D

End of message
",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render markdownBlockWithBoldVariable 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Hello Luke!
",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render markdownBlockWithTextWithTags 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Markdown text with <strong>tags</strong>
Should work.
",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render markdownBlockWithVariableWithTags 1`] = `
Array [
  Object {
    "text": Object {
      "text": "​*markdown*​ needs to support html <em>tags</em>.

As well as new lines

and new lines created by using &lt;br&gt;.
",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render quoteBlockWithLink 1`] = `
Array [
  Object {
    "text": Object {
      "text": "> Quote block with a <https://example.com|link>.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render templateBlock 1`] = `Array []`;

exports[`slack template blocks should render textBlock 1`] = `
Array [
  Object {
    "text": Object {
      "text": "This is a text block",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockHeader 1`] = `
Array [
  Object {
    "text": Object {
      "text": "You cannot hide forever, Luke.

Give yourself to the Dark Side.",
      "type": "plain_text",
    },
    "type": "header",
  },
]
`;

exports[`slack template blocks should render textBlockWhitespace 1`] = `
Array [
  Object {
    "text": Object {
      "text": "To trim or not to trim?  
",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithLink 1`] = `
Array [
  Object {
    "text": Object {
      "text": "<https://example.com|Luke>",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithLinkWithWhitespace 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Link<https://example.com|with whitespace>is fun",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithMarks 1`] = `
Array [
  Object {
    "text": Object {
      "text": "*You cannot hide forever,* *Luke.*

_Give yourself to the Dark Side._",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithVariable 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Hello Luke",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithVariableEscaping 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Escape variable values: Should escape &, <, >, \\", ', \`, and = characters.
Escape link values: <https://example.com/?a=1&b=2&c=3|Should escape &, <, >, \\", ', \`, and = characters>.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithVariableWithLineReturns 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Var with line returns: Line 1
Line 2

Line 4.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render textBlockWithVariableWithWindowsLineReturns 1`] = `
Array [
  Object {
    "text": Object {
      "text": "line 1
line 2

line 4",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;

exports[`slack template blocks should render welcomeBlocks 1`] = `
Array [
  Object {
    "text": Object {
      "text": "Welcome to Courier!",
      "type": "plain_text",
    },
    "type": "header",
  },
  Object {
    "text": Object {
      "text": "*These are content blocks.*",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "text": Object {
      "text": "• Feel free to drag and drop me around. 
• If you remove us don’t worry, We’ll simple move into the library under Channels for later use. 
• A block can be shared across channels! 
• When you are done, press Preview to send yourself a notification.
• Use curly brackets *{* and *}* to create {variables}.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  Object {
    "elements": Array [
      Object {
        "action_id": undefined,
        "text": Object {
          "emoji": true,
          "text": "Click to edit my properties!",
          "type": "plain_text",
        },
        "type": "button",
        "url": "https://www.courier.com",
      },
    ],
    "type": "actions",
  },
  Object {
    "text": Object {
      "text": "Cheers, 
*The Courier Team*",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;
